source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "emmintrin.h"
+#include "immintrin.h"
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <aws/common/common.h>
+
+/***** Decode logic *****/
+
+/*
+ * Decodes ranges of bytes in place
+ * For each byte of 'in' that is between lo and hi (inclusive), adds offset and _adds_ it to the corresponding offset in
+ * out.
+ */
+static inline __m256i translate_range(__m256i in, uint8_t lo, uint8_t hi, uint8_t offset) {
+    __m256i lovec = _mm256_set1_epi8(lo);
+    __m256i hivec = _mm256_set1_epi8(hi - lo);
+    __m256i offsetvec = _mm256_set1_epi8(offset);
+
+    __m256i tmp = _mm256_sub_epi8(in, lovec);
+    /*
+     * we'll use the unsigned min operator to do our comparison. Note that
+     * there's no unsigned compare as a comparison intrinsic.
+     */
+    __m256i mask = _mm256_min_epu8(tmp, hivec);
+    /* if mask = tmp, then keep that byte */
+    mask = _mm256_cmpeq_epi8(mask, tmp);
+
+    tmp = _mm256_add_epi8(tmp, offsetvec);
+    tmp &= mask;
+    return tmp;
+}
+
+/*
+ * For each 8-bit element in in, if the element equals match, add to the corresponding element in out the value decode.
+ */
+static inline __m256i translate_exact(__m256i in, uint8_t match, uint8_t decode) {
+    __m256i mask = _mm256_cmpeq_epi8(in, _mm256_set1_epi8(match));
+    return mask & _mm256_set1_epi8(decode);
+}
+
+/*
+ * Input: a pointer to a 256-bit vector of base64 characters
+ * The pointed-to-vector is replaced by a 256-bit vector of 6-bit decoded parts;
+ * on decode failure, returns 0, else returns 1 on success.
+ */
+static inline bool decode_vec(__m256i *in) {
+    __m256i tmp;
+
+    tmp = translate_range(*in, 'A', 'Z', 0 + 1);
+    tmp |= translate_range(*in, 'a', 'z', 26 + 1);
+    tmp |= translate_range(*in, '0', '9', 52 + 1);
+    tmp |= translate_exact(*in, '+', 62 + 1);
+    tmp |= translate_exact(*in, '/', 63 + 1);
```

Comment:
Don't use magic numbers. Doing arithmetic on `'/' - 'A'` is totally fine.

Reply:
This is a base64 encoding table, effectively. These numbers are the base64 indices that correspond with 'a', '0', '+', '/', etc. `'/' - 'A'` is incorrect as these ranges are not contiguous. Extracting constants will just obscure this mapping - which is easier to compare with the RFC, `translate_exact(*in, '/', BASE64_FORWARD_SLASH_INDEX + 1)` or `translate_exact(*in, '/', 63 + 1)`?

Reply:
Well, I have no way to compare to an RFC if I don't know about the RFC? I'm totally fine with keeping those numbers if you add a link to the RFC as a comment above the code. That would remove all "magic" from the numbers.